### PR TITLE
jk/new 3daxis attempt

### PR DESCRIPTION
**Pull request by jkrumbiegel**
- add 3d axis
- check axis with can_be_current_axis
- gridlines, ticks, fix inv eyepos
- axis3 as default 3d axis
- add ticks
- change to ax3 for fps as well

### CODE REVIEW

**Comment by Datseris** [comment]
I'm testing this locally on my own projects now. Thanks a lot, I am really desperate of unified look between 2D and 3D ;) .

This is the test code i used:
```
fig, ax3, surf = surface(1:100, 1:100, randn(100, 100))
display(fig)
angles = 0:0.01:2π
for angle in angles
    ax3.azimuth = angle
    sleep(0.001)
end
```
image



Which works really well. Two comments for now.

1. perhaps it is better to auto-scale the axes, if the user does not explicitly provide bounds, so that they all have the same size, instead of data size indicating a much smaller axis. This is a problem the old 3D axis also has. For example, 2D plotting does not do this by default: if I plot 1:100 versus 0.00001:0.0001:0.0001, I don't get a super-stretched line as an AXis, but a square.
2. Attempting to change GLmakie window size after the plot leads to the error:


```
julia> InexactError: UInt64(Inf)
Stacktrace:
 [1] UInt64 at .\float.jl:689 [inlined]
 [2] Timer(::Float64; interval::Float64) at .\asyncevent.jl:74
 [3] Timer at .\asyncevent.jl:72 [inlined]
 [4] sleep at .\asyncevent.jl:213 [inlined]
 [5] fps_renderloop(::GLMakie.Screen, ::Float64) at C:\Users\datse\.julia\packages\GLMakie\AiucZ\src\rendering.jl:32
 [6] renderloop(::GLMakie.Screen; framerate::Float64) at C:\Users\datse\.julia\packages\GLMakie\AiucZ\src\rendering.jl:48
 [7] renderloop(::GLMakie.Screen) at C:\Users\datse\.julia\packages\GLMakie\AiucZ\src\rendering.jl:41
 [8] (::GLMakie.var"#51#53"{GLMakie.Screen})() at .\task.jl:356
```

**Comment by jkrumbiegel** [comment]
can you try setting `ax.data_aspect = :equal` for the scaling thing, it should do the different scaling behavior

**Comment by Datseris** [comment]
It works:

image


I would still argue that this should be the default behavior in 3D

**Comment by jkrumbiegel** [comment]
sure I was just testing the other mode last

**Comment by Datseris** [comment]
Another problem I'm having is interactively changing the 3D axis limits for animation.
I have:
```
julia> ax3.limits
Observable{GeometryBasics.HyperRectangle{3,Float32}} with 19 listeners. Value:  
GeometryBasics.HyperRectangle{3,Float32}(Float32[0.084998, -0.121002, 0.074998], Float32[0.085002, -0.120998, 0.075002])
```

and I create an `FRect3D` to pass to `ax3.limits`. Doing

`ax3.limits = FRect3D(orig, wid)` gives error 
```
ERROR: TypeError: in setfield!, expected Observable{GeometryBasics.HyperRectangle{3,Float32}}, got a value of type GeometryBasics.HyperRectangle{3,Float32}     
Stacktrace:
 [1] setproperty!(::Axis3, ::Symbol, ::GeometryBasics.HyperRectangle{3,Float32}) at C:\Users\datse\.julia\packages\AbstractPlotting\cKGdX\src\makielayout\layoutables.jl:114
```

while doing `ax3.limits[] = FRect...` gives 
```
ERROR: ArgumentError: reducing over an empty collection is not allowed
Stacktrace:
 [1] _empty_reduce_error() at .\reduce.jl:299
 [2] reduce_empty(::Function, ::Type{Float64}) at .\reduce.jl:309
 [3] mapreduce_empty(::typeof(identity), ::Function, ::Type{T} where T) at .\reduce.jl:343
 [4] reduce_empty(::Base.MappingRF{typeof(identity),typeof(min)}, ::Type{Float64}) at .\reduce.jl:329
 [5] reduce_empty_iter at .\reduce.jl:355 [inlined]
 [6] mapreduce_empty_iter(::Function, ::Function, ::Array{Float64,1}, ::Base.HasEltype) at .\reduce.jl:351
 [7] _mapreduce(::typeof(identity), ::typeof(min), ::IndexLinear, ::Array{Float64,1}) at .\reduce.jl:400
 [8] _mapreduce_dim at .\reducedim.jl:318 [inlined]
 [9] #mapreduce#620 at .\reducedim.jl:310 [inlined]
 [10] mapreduce at .\reducedim.jl:310 [inlined]
 [11] _minimum at .\reducedim.jl:727 [inlined]
 [12] _minimum at .\reducedim.jl:726 [inlined]
 [13] #minimum#633 at .\reducedim.jl:722 [inlined]
 [14] minimum at .\reducedim.jl:722 [inlined]
 [15] locateticks(::Float32, ::Float32, ::Int64, ::Bool, ::Int64) at C:\Users\datse\.julia\packages\AbstractPlotting\cKGdX\src\makielayout\ticklocators\linear.jl:144
 [16] locateticks at C:\Users\datse\.julia\packages\AbstractPlotting\cKGdX\src\makielayout\ticklocators\linear.jl:77 [inlined]
 [17] get_tickvalues at C:\Users\datse\.julia\packages\AbstractPlotting\cKGdX\src\makielayout\lineaxis.jl:451 [inlined]
 [18] #270 at C:\Users\datse\.julia\packages\AbstractPlotting\cKGdX\src\makielayout\layoutables\axis3d.jl:46 [inlined]
 [19] MapUpdater at C:\Users\datse\.julia\packages\Observables\Yf3xU\src\Observables.jl:331 [inlined]
 [20] (::Observables.OnUpdate{Observables.MapUpdater{AbstractPlotting.MakieLayout.var"#270#279",Array{Float64,1}},Tuple{Observable{GeometryBasics.HyperRectangle{3,Float32}}}})(::GeometryBasics.HyperRectangle{3,Float32}) at C:\Users\datse\.julia\packages\Observables\Yf3xU\src\Observables.jl:298
 [21] setindex!(::Observable{GeometryBasics.HyperRectangle{3,Float32}}, ::GeometryBasics.HyperRectangle{3,Float32}; notify::Observables.var"#9#11") at C:\Users\datse\.julia\packages\Observables\Yf3xU\src\Observables.jl:205
 [22] setindex!(::Observable{GeometryBasics.HyperRectangle{3,Float32}}, ::GeometryBasics.HyperRectangle{3,Float32}) at C:\Users\datse\.julia\packages\Observables\Yf3xU\src\Observables.jl:201
```

P.s.: If you don't like me reporting problems, let me know

**Comment by Datseris** [comment]
For large tick labels, they cover the axis labels:
image

File: docs/src/makielayout/axis3.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+## Data aspects and view mode
+
+The attributes `data_aspect` and `viewmode` both influence the apparent relative scaling of the three axes.
+
+### `data_aspect`
+
+The `data_aspect` changes how long each axis is relative to the other two.
+
+If you set it to `:data`, the axes will be scaled according to their lengths in data space.
+The visual result is that objects with known real-world dimensions look correct and not squished or
```
Comment (Datseris) [inline docs/src/makielayout/axis3.md:12]:
```suggestion
The visual result is that objects with known real-world dimensions look correct and not squished.
```

**Pull request merged**